python/lsst/ctrl/mpexec/cli/cmd/commands.py

```diff
@@ -75,7 +76,11 @@ def build(ctx, **kwargs):
 
     This does not require input data to be specified.
     """
-    _doBuild(ctx, **kwargs)
+    try:
+        _doBuild(ctx, **kwargs)
+    except Exception:
+        traceback.print_exc()
+        ctx.exit(1)
```

Comment:
Interesting. I did a bit of an experiment before and after because I wasn't initially clear what was happening:

```
$ butler create REPO
Traceback (most recent call last):
  File "/Users/timj/work/lsstsw3/build/daf_butler_test/bin/butler", line 28, in <module>
    sys.exit(main())
  File "/Users/timj/work/lsstsw3/build/daf_butler_test/python/lsst/daf/butler/cli/butler.py", line 334, in main
    return cli()
  File "/Users/timj/work/lsstsw3/miniconda/envs/lsst-scipipe-0.6.0/lib/python3.8/site-packages/click/core.py", line 1137, in __call__
    return self.main(*args, **kwargs)
  File "/Users/timj/work/lsstsw3/miniconda/envs/lsst-scipipe-0.6.0/lib/python3.8/site-packages/click/core.py", line 1062, in main
    rv = self.invoke(ctx)
  File "/Users/timj/work/lsstsw3/miniconda/envs/lsst-scipipe-0.6.0/lib/python3.8/site-packages/click/core.py", line 1668, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/timj/work/lsstsw3/miniconda/envs/lsst-scipipe-0.6.0/lib/python3.8/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/timj/work/lsstsw3/miniconda/envs/lsst-scipipe-0.6.0/lib/python3.8/site-packages/click/core.py", line 763, in invoke
    return __callback(*args, **kwargs)
  File "/Users/timj/work/lsstsw3/build/daf_butler_test/python/lsst/daf/butler/cli/cmd/commands.py", line 117, in create
    script.createRepo(*args, **kwargs)
  File "/Users/timj/work/lsstsw3/build/daf_butler_test/python/lsst/daf/butler/script/createRepo.py", line 48, in createRepo
    raise RuntimeError("Oops")
RuntimeError: Oops
(lsst-scipipe-0.6.0) thrylos:daf_butler_test (tickets/DM-31696) $ butler create REPO
Traceback (most recent call last):
  File "/Users/timj/work/lsstsw3/build/daf_butler_test/python/lsst/daf/butler/cli/cmd/commands.py", line 119, in create
    script.createRepo(*args, **kwargs)
  File "/Users/timj/work/lsstsw3/build/daf_butler_test/python/lsst/daf/butler/script/createRepo.py", line 48, in createRepo
    raise RuntimeError("Oops")
RuntimeError: Oops
```
So yes, it does strip the "we don't care about this" code from the exception traceback. I assume we should fix bps and butler commands the same way. I had a feeling that @n8pease had done some exception catching trickery before but I can't found it now so maybe I was imagining it.

Reply:
I can update butler cli to do the same on this ticket if you prefer (and probably bps, but I never looked at that code).

Reply:
Thanks. BPS looks exactly like pipetask in that it uses click and is set up in the same way.

Reply:
Can't this instead be handled at a level higher at https://github.com/lsst/daf_butler/blob/master/python/lsst/daf/butler/cli/butler.py#L334 by stripping levels from the top instead of exiting out at the bottom?  That would mean one change instead of a change for every subcommand.

Reply:
You would need to know how many stack levels you need to strip, which is sort of non-trivial. I am trying now to simplify things by defining a re-usable decorator which needs just one line change for each method  (but still need to update every method).

Reply:
I was thinking it would be relatively easy to strip until you get beyond `click`.

Reply:
Exactly. I had a similar problem in some astropy code working out what stacklevel to use for warning message to make the message show it came from user code (so I walked the stack). We know that click is at the top level and finding where click stops seems like a reasonable thing to do.

Reply:
That feels a bit fragile as it depends on assumption that click does not call any other library before it lands in our code. Let me show you my decorator code and then you decide if that is too complicated.

Reply:
OK, here is how I'm suggesting to do it [with a decorator](https://github.com/lsst/daf_butler/pull/572/commits/7681e57d883b660b8b1b04db5731309a6282ef79)

Reply:
Is it not possible to put that decorator once in the `main` definition that @ktlim pointed to above?

Reply:
That will show click in traceback, to exclude click I'd need to do something much more non-trivial and fragile.

Reply:
OK, another [attempt](https://github.com/lsst/daf_butler/pull/572/commits/906c6aa405397e3427f4310b830615d7eb4b5863) that does not require a per-method decorators, this depends on our overriding of `click.Command` so I can (ab)use it to wrap a callback method with the same decorator. This does not always work perfectly but in those cases I can keep explicit decorator (this only concerns three methods in ctrl_mpexec).